admin/tools/provision/cloudmanager.py

```diff
@@ -0,0 +1,311 @@
+        """
+        logging.info('Manage ssh keys: {}'.format(self.key))
+        if self.nova.keypairs.findall(name=self.key):
+            logging.debug('Remove previous ssh keys')
+            self.nova.keypairs.delete(key=self.key)
+
+        with open(os.path.expanduser(self.key_filename + ".pub")) as fpubkey:
+            self.nova.keypairs.create(name=self.key, public_key=fpubkey.read())
+
+    def get_floating_ip(self):
+        """
+        Allocate floating ip address to project
+        """
+        i = 0
+        floating_ips = self.nova.floating_ips.list()
+        floating_ip = None
+        floating_ip_pool = self.nova.floating_ip_pools.list()[0].name
+
+        # Check for available public ip address in project
+        while i < len(floating_ips) and floating_ip is None:
+            if floating_ips[i].instance_id is None:
+                floating_ip = floating_ips[i]
+                logging.debug('Available floating ip found {}'.format(floating_ip))
+            i += 1
+
+        # Check for available public ip adress in ext-net pool
+        if floating_ip is None:
+            try:
+                logging.debug("Use floating ip pool: {}".format(floating_ip_pool))
+                floating_ip = self.nova.floating_ips.create(floating_ip_pool)
+            except novaclient.exceptions.Forbidden as exc:
+                logging.fatal("Unable to retrieve public IP: {0}".format(exc))
+                sys.exit(1)
+
+        return floating_ip
+
+    def print_ssh_config(self, instances, floating_ip):
+        """
+        Print ssh client configuration to file
+        """
+        # ssh config
+        ssh_config_tpl = '''
+        Host {host}
+        HostName {fixed_ip}
+        User qserv
+        Port 22
+        StrictHostKeyChecking no
+        UserKnownHostsFile /dev/null
+        PasswordAuthentication no
+        ProxyCommand ssh -i {key_filename} -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -W %h:%p qserv@{floating_ip}
+        IdentityFile {key_filename}
+        IdentitiesOnly yes
+        LogLevel FATAL
+        '''
+        ssh_config_extract = ""
+        for instance in instances:
+            fixed_ip = instance.networks[self.network_name][0]
+            ssh_config_extract += ssh_config_tpl.format(host=instance.name,
+                                                        fixed_ip=fixed_ip,
+                                                        floating_ip=floating_ip.ip,
+                                                        key_filename=self.key_filename)
+        logging.debug("Create SSH client config ")
+
+        f = open("ssh_config", "w")
+        f.write(ssh_config_extract)
+        f.close()
+
+    def check_ssh_up(self, instances):
+        """
+        Check if the ssh service started
+        """
+        for instance in instances:
+            cmd = ['ssh', '-t', '-F', './ssh_config', instance.name, 'true']
+            success = False
+            while not success:
```

Comment:
This looks like an infinite loop in case something wrong happens for a particular host. Maybe some timeout could be useful here (but then you need to decide what to do in case of failure)?